flips/20201125-bip-44-multi-account.md

```diff
@@ -0,0 +1,323 @@
+44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) (SLIP 44).
+
+#### Account
+
+Account is the index of a distinct account address owned by the wallet user. No
+two account indices should be used to generate child public keys that are
+assigned to the same Flow account address. Furthermore, a single account index
+should not be used to generate keys for more than one account address.
+
+The motivation for these accounts is the same set forth in BIP 44:
+
+> Users can use these accounts to organize the funds in the same fashion as bank
+> accounts; for donation purposes (where all addresses are considered public),
+> for saving purposes, for common expenses etc.
+
+When generating a key pair to be added on a new account, use the smallest 
+account index that has not already been used to generate keys for an existing account.
+
+Unlike [BIP 44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account),
+new accounts should be able to be created even if previously created accounts
+have no transaction history.
+
+#### Change
+
+Unlike Bitcoin, Flow does not utilize change addresses. Change is always the
+constant 0.
+
+#### Index
+
+In BIP 44, each index generates a public key that is used to derive a unique
+address for an account. Although Flow does not derive account addresses from
+public keys, index can still be used to generate multiple public keys for a
+single account.
+
+When applying BIP 44 to Flow, each subsequent value for index should be used to
+generate a new key pair for the same account, starting from 0 and incrementing
+by 1.
+
+This index should not be confused with the on-chain account key index. It must
+not be assumed that this index will reflect the on-chain account key index for
+any key.
+
+When generating a new key pair for an account, the key index to use in the path to 
+generate the new key should be the smallest unused key index available which 
+has not yet been used to generate a key on the account.
+
+### Legacy Path
+
+There exist Flow accounts with keys assigned to them which have been generated
+by wallets using the legacy path.
+
+The legacy path has been previously used by FCL-Ledger-Web and is what has been used to
+generate keys for Ledger Flow users. As far as we are aware, this is the only
+application which has used the legacy path.
+
+The legacy path is _deprecated_ by this FLIP, and should not be used to
+generate new keys.
+
+The legacy path is:
+
+```
+m / 44' / 1' / 769 / 0 / 0
+```
+
+All keys generated with the legacy path as of today (Dec 2021) have been done so using the
+NIST P-256 elliptic curve.
+
+### Account Discovery
+
+Wallet developers should use a modified version of the original account
+discovery procedure [described in BIP
+44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery).
+
+Modified constants from BIP 44 include:
+
+The key index gap limit should be `5`.
+The account index gap limit should be `5`.
```

Comment:
Why `5` ? Seems like a lot. Does this mean a wallet should derive 2(two curves) * 5(key_gap) * 5(account_gap) keys in order to say that a wallet is not used? In other words, should the wallet derive 10 keys, just to say the account is not used? Am I getting this wrong? What is the use case for such a big gap limit?

Since keys and accounts should be used in order (meaning the first unused should be picked when creating new account/key) why not gap limit `1`?.

Reply:
Hey @PeterBenc , after consulting with the team, we've concluded that the key gap limit and account gap limit parameters should be rather flexible. Developers who wish to use a smaller key or address gap limit should be able to do so if they choose - developers should use their best judgement when choosing a key gap limit and account gap limit.

Reply:
Thanks for considering this, makes sense, at least for now we are gonna use account_gap_limit 1, and check only the first key (so I guess key_gap_limit 0) since right now, basically nobody is deriving these keys anyway and we can increase the gap limit later if needed. 

Just genuinely interested, I suppose the 5 and 2 (for key account gap respectively) should have some recommending purpose? Since it's supposed to be flexible, what is the reason for defining these?

Reply:
@PeterBenc
In BTC for example addresses are bound to keys, so if you have HD wallet you can have unlimited addresses.

so lets say my wallet pre generates N addresses for me, and I use the Nth to receive some coins. 

After I restore my wallet to another computer, you need to search N addresses (gap = N ) otherwise I see my funds are missing and panic 

i dont know why we need in Flow , I need to reread to FLIP